File: algorithms/java/src/test/java/leetcode/LC300LongestIncreasingSubsequence.java
package leetcode.lc300_longest_increasing_subsequence;

import org.junit.jupiter.api.Test;


import java.util.Arrays;

import static org.junit.jupiter.api.Assertions.assertEquals;

/**
 * https://leetcode.com/problems/longest-increasing-subsequence/
 * #medium
 */
public final class LC300LongestIncreasingSubsequence {
    public final class Solution {
        public int lengthOfLIS(final int[] nums) {
            if (nums == null || nums.length == 0) return 0;
            int[] lis = new int[nums.length];
            Arrays.fill(lis, 1);
            for (int i = 1; i < nums.length; i++) {
                for (int j = 0; j < i; j++) {
                    if (nums[i] > nums[j]) {
                        lis[i] = Math.max(lis[i], lis[j] + 1);
                    }
                }
            }
            int max = Integer.MIN_VALUE;
            for (int value : lis) {
                max = Math.max(max, value);
            }
            return max;
        }
    }

    @Test
    public void test_example() throws Exception {
        int[] nums = {10, 9, 2, 5, 3, 7, 101, 18};
        assertEquals(4, new Solution().lengthOfLIS(nums));
    }
}

File: algorithms/code/leetcode/lc300_longest_increasing_subsequence/lc300_longest_increasing_subsequence.py
# https://leetcode.com/problems/longest-increasing-subsequence/

import unittest


class Solution:
    def lengthOfLIS(self, nums: list[int]) -> int:
        if len(nums) == 0:
            return 0
        lis = [1] * len(nums)
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[i] > nums[j]:
                    lis[i] = max(lis[i], lis[j] + 1)
        maximum = max(lis)
        return maximum


class TestCode(unittest.TestCase):
    def test_example(self) -> None:
        nums = [10, 9, 2, 5, 3, 7, 101, 18]
        self.assertEqual(4, Solution().lengthOfLIS(nums))

    def test_nothing(self) -> None:
        self.assertEqual(0, Solution().lengthOfLIS([]))
